### PR TITLE
Add additionalInfo to CloudError

### PR DESCRIPTION
**Pull request by j5lim**
## Description

'additionalInfo', a new property added to ARM error contract, contains additional error information. For example, it will contain detailed policy violation error information when the request to ARM failed due to violation of Azure policies.

The changes are to allow deserializing additionalInfo in ARM error based on the error type.
---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by dsgouda** [comment]
@shahabhijeet we need to take a look at this

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure.Tests/CloudErrorJsonConverterTest.cs [hunk 1]
```diff
@@ -26,6 +26,14 @@ public void TestCloudErrorDeserialization()
                             ""target"": ""$search"",
                             ""message"": ""$search query option not supported""
                         }
+                        ],
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure.Tests/CloudErrorJsonConverterTest.cs]: Can you add a scenario that mimics legacy error responses (essentially that do not have the additive changes)
This is a runtime component that is being shared with all services for every API version released and this change is especially crucial when we talk about Multi-API packages.

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure.Tests/CloudErrorJsonConverterTest.cs [hunk 2]
```diff
@@ -93,5 +104,82 @@ public void TestEmptyCloudErrorDeserialization()
             Assert.NotNull(cloudError);
             Assert.Null(cloudError.Code);
         }
+
+        [Fact]
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure.Tests/CloudErrorJsonConverterTest.cs]: @j5lim this test is good, but after the change requested of removing actual violation type, you might have to rewrite the asserts

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/CloudError.cs [hunk 3]
```diff
@@ -37,5 +38,11 @@ public CloudError()
         /// Gets or sets details for the error.
         /// </summary>
         public IList<CloudError> Details { get; private set; }
+
+        /// <summary>
+        /// Gets or sets additional error info.
+        /// </summary>
+        [JsonProperty(ItemConverterType = typeof(TypedErrorInfoJsonConverter))]
+        public IList<TypedErrorInfo> AdditionalInfo { get; set; }
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/CloudError.cs]: @j5lim this change has a potential to introduce runtime exceptions for legacy AP versions for services.
This change will need to thoroughly tested for all services (including past API versions)

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.csproj [hunk 4]
```diff
@@ -4,7 +4,7 @@
     <PackageId>Microsoft.Rest.ClientRuntime.Azure</PackageId>
     <Description>Provides common error handling, tracing, and HTTP/REST-based pipeline manipulation.</Description>
     <AssemblyTitle>Client Runtime for Microsoft Azure Libraries</AssemblyTitle>
-    <Version>3.3.12</Version>
+    <Version>3.3.13</Version>
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.csproj]: @j5lim can you revert this change, I will make sure the version is updated before publishing.
Version bump will not be accepted along with this PR.

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -9,7 +9,7 @@
 [assembly: AssemblyTitle("Microsoft Rest Azure Client Runtime")]
 [assembly: AssemblyDescription("Client infrastructure for Azure client libraries.")]
 [assembly: AssemblyVersion("3.0.0.0")]
-[assembly: AssemblyFileVersion("3.3.12.0")]
+[assembly: AssemblyFileVersion("3.3.13.0")]
```
Comment (shahabhijeet) [inline src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/Properties/AssemblyInfo.cs]: @j5lim can you revert this change. The version number will be updated once this is ready to be published.

**Review by shahabhijeet — CHANGES_REQUESTED** [review]

File: src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/CloudError.cs (hunk 3 above)
Comment (j5lim) [reply src/SdkCommon/ClientRuntime.Azure/ClientRuntime.Azure/CloudError.cs]: I think the tests already cover that case (the response missing additionalInfo property, which is optional). Do you think it is not enough?

**Comment by shahabhijeet** [comment]
@j5lim closing this in lieu of the following PR
https://github.com/Azure/azure-sdk-for-net/pull/4635, plus this PR has merge conflicts
